packages/toolkit/src/login/webview/vue/backend.ts

```diff
@@ -0,0 +1,166 @@
+                    id: 'invalidGrantException',
+                    text: 'Permissions for this service may not be enabled by your SSO Admin, or the selected region may not be supported.',
+                }
+            }
+
+            if (
+                e instanceof ToolkitError &&
+                (e.code === trustedDomainCancellation || e.cause?.name === trustedDomainCancellation)
+            ) {
+                return {
+                    id: 'trustedDomainCancellation',
+                    text: `Must 'Open' or 'Configure Trusted Domains', unless you cancelled.`,
+                }
+            }
+
+            const invalidRequestException = 'InvalidRequestException'
+            if (
+                (e instanceof Error && e.name === invalidRequestException) ||
+                (e instanceof ToolkitError && e.cause?.name === invalidRequestException)
+            ) {
+                return { id: 'badStartUrl', text: `Connection failed. Please verify your start URL.` }
+            }
+
+            // If SSO setup fails we want to be able to show the user an error in the UI, due to this we cannot
+            // throw an error here. So instead this will additionally show an error message that provides more
+            // detailed information.
+            handleWebviewError(e, this.id, methodName)
+
+            return { id: 'defaultFailure', text: 'Failed to setup.' }
+        }
+    }
+
+    async startBuilderIdSetup(app: string): Promise<AuthError | undefined> {
+        return this.ssoSetup('startCodeWhispererBuilderIdSetup', async () => {
+            if (app === 'AMAZONQ') {
+                await awsIdSignIn()
+                AuthUtil.instance.hasAlreadySeenMigrationAuthScreen = true
+                await vscode.window.showInformationMessage('AmazonQ: Successfully connected to AWS Builder ID')
+            } else {
+                // no builder id in toolkit
+            }
+        })
+    }
+
+    async startEnterpriseSetup(startUrl: string, region: string, app: string): Promise<AuthError | undefined> {
+        if (app === 'AMAZONQ') {
+            return this.ssoSetup('startCodeWhispererEnterpriseSetup', async () => {
+                await connectToEnterpriseSso(startUrl, region)
+                AuthUtil.instance.hasAlreadySeenMigrationAuthScreen = true
+                void vscode.window.showInformationMessage('AmazonQ: Successfully connected to AWS IAM Identity Center')
+            })
+        } else {
+            return this.ssoSetup('createIdentityCenterConnection', async () => {
+                const ssoProfile = createSsoProfile(startUrl, region)
+                const conn = await Auth.instance.createConnection(ssoProfile)
+                await Auth.instance.useConnection(conn)
+                void vscode.window.showInformationMessage('Toolkit: Successfully connected to AWS IAM Identity Center')
+                void this.showResourceExplorer()
+            })
+        }
+    }
```

Comment:
this class and especially these functions have hardcoded options that you can pass in. this will become messy and difficult to maintain if we ever add new functionality that requires its own auth.

for an initial prototype, this is fine. But we should think about how we can make this more dynamic later for additional usecases without a bunch of specialized if statements.

Reply:
I am aware of this. I have yet found an very effective way to make it very clean.

Reply:
This is fixed in latest commit